### ISSUE TITLE
GE doesn't load commits on demand

### ISSUE THREAD

**Issue by RussKie** [gitextensions/gitextensions#5114]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
If GE is configured to load less commits on start up then there are commits in the repository then the remaining commits won't be loaded at all.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
* Reduce the number of loaded commits on startup from 100,000 to 100 or 1,000
![image](https://user-images.githubusercontent.com/4403806/41811411-78115266-7752-11e8-8748-d8e8283283bb.png)
* Close/open GE
* Scroll down and observe no commits past the cut-out point
![image](https://user-images.githubusercontent.com/4403806/41811420-af777a0a-7752-11e8-9d2b-63cb3fb1e9ba.png)

**What is the expected behavior?**
Commits are loaded on demand

**Environment you encounter the issue:**

- GitExtensions version: all
- GIT version: 2.16.2

**Did this work in previous version of GitExtensions (which)?**
I don't believe so

I'm pretty sure #4982 is related.

**Comment by RussKie** [comment]
@spdr870 any thoughts as to how this may be addressed? 
Not expecting anything immediate, more a discussion at this point

This relates to #5724

**Comment by spdr870** [comment]
In 3.0 it at least looks better, the lines continue.

This is quite easy now. Just feed more commits to the graph. Just start a new git.exe log command that continues from that commit. There are some issues to keep in mind:
- the quick search and go to commit will only look in the loaded commits
- the scroll bar won't be there, unless the commits are counted first
- if the history changed in between loading parts, you might get strange results

To me, this is not really a bug but a feature. I hardly ever need to load commits that old. I changed the limit to 200.000 myself. In version 3.00 its fast enough.

Nice fact; very early versions of gitextensions actually had this feature. It didn't load extra commits, but reloaded the graph with 2x the number of commits. Now the graph code should be able to handle it.

We could also do an easier solution. If the bottom is reached, show a button and text with 'log is truncated. Click here to load the entire log'. It could then reload. Its slower, but I do not expect it to be used a lot. Personally i do not really like those magic infinite scroll grids if loading requires heavy  processing power.

**Comment by RussKie** [comment]
I like the "load more" button idea.

**Comment by vbjay** [comment]
Instead of a button, if truncated and reached bottom scroll position load x
window more commits. To know if truncated will have to use a variable to
hold loaded count and init it to app settings on each module change. If we
load more, increase by commit count loaded. If scroll to bottom and log
returns 0 commits set a fully loaded flag. We have a loading notification
in the corner.

**Comment by spdr870** [comment]
@vbjay You don't have to use a variable. You can just ask the RevisionGraph, it knows. When _nodes.Count < _nodersIndexedById.Count the log is truncated..

Problem is that its hard to load the next x commits. How to know where to start?

**Comment by vbjay** [comment]
I'll have to check but wouldn't git log with bottom Sha  --all -n... Load logs from that point back?

**Comment by vbjay** [comment]
Ok.  Set limit to 50 commits.  Scroll to bottom commit of a repo that is truncated.  Copy sha.
![image](https://user-images.githubusercontent.com/3629489/48228575-937d3280-e373-11e8-80fa-05ab00af77b5.png)
d8788a8e92fc0ec76af07e2bef0f6df397ccb572
```bash
git log --all -n 50 d8788a8e92fc0ec76af07e2bef0f6df397ccb572
```

**Comment by spdr870** [comment]
I have something working with a page size of 10k commits. It breaks searching and goto commit, but that is fixable. I do not really like how the scrollbar jumps, it makes scrolling harder. With a large pagesize it is probably ok. Initial memory usage is very low and the app is more resposive on large repos. With roughly 10 lines of code its a small change.
![loadonscroll](https://user-images.githubusercontent.com/38675/49023300-3b2d9980-f197-11e8-8c16-8e787a33093c.gif)

**Comment by RussKie** [comment]
Nice!

**Comment by RussKie** [comment]
@spdr870 I don't believe you have pushed your changes, and the app isn't loading commits on demand.

Can you please send the changes in?
